Add nullsafety and clean repo and code

Wikiwix opened the pull request:
## Functional changes

* Fix an issue that lets the `pmap` fail if the results need to be sorted.
* Add an optional `preserveOrder` option to ignore order and faster get new values from the stream. (default is `true`, making the API surface downward compatible)
* Create an `mapParallel` extension method for `Iterable`s.
* Test whether all the iterable elements are successfully processed. Otherwise throw a `StateError`.
* Migrate to nullsafety.

## Repo changes

*  Add `.gitignore`
* Simplify `example` folder
* Add and fulfil `pedantic` lints

Wikiwix commented:
@gaaclarke Should I add a GitHub action as well to automatically test changes (unsure in how far that impacts your personal quota on this)?

gaaclarke commented:
@Wikiwix Thanks, I never setup github actions for my personal account.  Let's wait to hook up CI for now since the package isn't that popular.  I can look into it in the future if need be.

gaaclarke commented on README.md:
```diff
@@ -10,13 +10,11 @@ help parallelize expensive operations.
 import 'package:pmap/pmap.dart';
 
-int square(int x) => x * x;
+int mapper(int x) => x * x;
```
I would prefer to keep this square.  I think it is more clear to name what it does versus what it is used for.

Wikiwix replied:
I absolutely agree. I duplicated the example which used the `mapper` name. So I will just change both to `square` now.

gaaclarke commented:
I'm sorry, I don't have time to review this right now.  This is going to take a bit of time to review.  What I'd do is make sure that `dartanalyzer` doesn't give any linter or errors for the code.  I tried checking it out but dartanalyzer was tripping up for me on null safety.  If you don't hear back from me in a week, feel free to ping me, you won't be bothering me.

Wikiwix commented:
I completely understand. I am very sorry for the large PR, but what started as *just adding nullsafety* was getting out of hand quickly. (The initial code would have needed a lot of `?`s and especially `!`s, which I did not feel comfortable with. And then I experienced the bug (not knowing the soruce yet, which was when I decided to clean the code and make it more immutable.

If you mean

> What I'd do is make sure that dartanalyzer doesn't give any linter or errors for the code.

to be a suggestion to me: I do not know about `dartanalyzer` – which is no longer available in 2.12.0 (perhaps you activated some (old) version with `dart pub global` which is not on the 2.12.0 version yet?), but `dart analyze` gives

```bash
$ dart analyze --fatal-infos
Analyzing pmap...
No issues found!
```

In fact you can have a look here: https://github.com/Wikiwix/pmap/runs/2063750773

gaaclarke commented on lib/pmap.dart:
```diff
@@ -1,109 +1,178 @@
   return controller.stream;
 }
+
+extension PMapIterable<T> on Iterable<T> {
```
This is a good idea, nice!

gaaclarke commented on lib/pmap.dart:
```diff
@@ -1,109 +1,178 @@
+    return hasNext;
+  }
+
+// The number of already emitted elements
```
nit: leading spaces here and a couple lines down.

Wikiwix replied:
Thanks, I trusted in `dartfmt`, but for comments it was not strict enough.

gaaclarke commented on lib/pmap.dart:
```diff
@@ -1,109 +1,178 @@
-              receiveCount++;
-              index = buffer.indexWhere((x) => x.index == receiveCount);
+            final enumeratedResult = result as MapEntry<int, U>;
+            if (preserveOrder) {
```
nit: we could get better performance if we pulled out this boolean check so that it doesn't happen at each event.

Wikiwix replied:
I looked into this and I failed to do that without plain duplicating all the code in which case I would argue that the overhead is not worth this small gain.

If you have a suggestion on doing this more elegantly I am all ears!

What I did improve though, is making the `buffer` `late` so it only gets initialised in case it is used (my first time benefiting from the new keyword 🎉).

gaaclarke commented on test/pmap_test.dart:
```diff
@@ -3,24 +3,39 @@ import 'package:test/test.dart';
-    List<int> pmapped = await pmap(input, _square).toList();
-    expect(pmapped, equals(mapped));
-  });
+  final input = Iterable<int>.generate(1000);
```
Can you add a test that hits the `allElementsPublished = false` case?  Maybe throw an exception if a certain number input is given.

Wikiwix replied:
That was quite the oversight.

Unfortunately the result of trying to add it is a little sad:

* Throwing for some values just returns the error to the main thread (and halts the calculation, but does not exit)
* `exit`ing will kill the whole process not only the Isolate
* killing the Isolate results in the same issue as throwing – meaning the calculation halts but does not exit

If you want to try for yourself:

```dart
import 'dart:io';
import 'dart:isolate';

import 'package:pmap/pmap.dart';

int _squareWithIssues(int x) => x == 53
    ? () {
        throw Exception('test');
        // exit(1);
        // Isolate.current.kill();
        return -1;
      }()
    : x * x;

void main() async {
  final foo = Iterable<int>.generate(100000);
  foo.mapParallel(_squareWithIssues, parallel: 2).listen(print);
}
```
**My suggestion would be to remove the test logic from `pmap` for now, merge and release the update, and improve error handling afterwards. Do you agree?**

* The current PR is already really big, 
* the old version did not handle that case either
* the old version lacks null-safety
* the old version has the issue with ordering, which is quite toxic (as it silently drops values without errors)

(If we add proper error handling I personally feel like a `1.0.0` release is earned :D)

gaaclarke commented on lib/pmap.dart:
```diff
@@ -1,109 +1,178 @@
+  }
 }
 
+/// This [_Processor] is given to the spawned isolates as the work basis.
```
How about:
/// An object that represents all the data an isolate needs to perform its work.

gaaclarke commented on lib/pmap.dart:
```diff
@@ -1,109 +1,178 @@
+  _Processor(this.mapper, this.sendPort);
 }
 
+/// The data to manage a spawned [Isolate]
```
How about:
/// An isolate that will execute the `mapper` function on all items it receives via the `receivePort`.

gaaclarke reviewed (changes requested):
Thanks, this is a great PR.  The only thing it really needs is a unit test that exercises `!allElementesPublished` code.

gaaclarke commented:
@Wikiwix Also, did you know about Dart's work to make spawning isolates faster / less memory intensive?  If you are using pmap in production you might want to check out https://github.com/dart-lang/sdk/issues/36097  I enabled it for Flutter, I think it was just a command line flag.

Wikiwix commented:
@gaaclarke ~~I hope~~ I addressed all your comments now (either with a question, suggestion, or resolving it with the latest commit).

Thanks for the detailed review :) 👍🏻.

For the `!allElementesPublished` code/test I would await your response and then, if you agree with me, add a commit removing it (for now).

gaaclarke reviewed (approved):
I understand, you don't have to fix the error case.  I'll check it out when I get some time.  I just saw that it was missing from the tests.  This PR has already done a lot of good.  Thanks again for the contribution.

gaaclarke commented:
(I accidentally merged instead of rebased, so I rebased and forced pushed it after the merge, commit 85fc287e0904b855593f38fbdb92f388e3d9cbdf)
